### PR TITLE
Separate proof of assets and address ownership in the contract

### PR DESCRIPTION
**Pull request by alxkzmn**

### CODE REVIEW

File: contracts/src/Summa.sol [hunk 1]
```diff
@@ -18,39 +18,68 @@ interface Verifier {
...
+
+        emit ExchangeAddressesSubmitted(cexAddresses);
+    }
+
     /**
      * @dev Submit proof of solvency for a CEX
-     * @param exchangeId The ID of the exchange
-     * @param cexAddresses The CEX addresses
-     * @param cexSignatures The signatures of a message "Summa proof of solvency for {exchangeId}" for each CEX address
      * @param erc20ContractAddresses The addresses of the ERC20 token contracts that the CEX holds (e.g., USDT, USDC, DAI)
```
[the remark this answers is not in the document]
Comment (alxkzmn) [reply contracts/src/Summa.sol:72]: Sure, let's do it

File: contracts/src/Summa.sol [hunk 2]
```diff
@@ -18,39 +18,68 @@ interface Verifier {
...
      * @param balancesToProve The balances to prove. ETH balance should be the first element, followed by ERC20 balances in the order of erc20ContractAddresses
```
[the remark this answers is not in the document]
Comment (alxkzmn) [reply contracts/src/Summa.sol:73]: Ok!

File: contracts/src/Summa.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (alxkzmn) [reply contracts/src/Summa.sol:72]: Right now for each of the addresses submitted in `submitProofOfAccountOwnership` we calculate total ETH balance and total ERC20 balance (per token) - we're assuming that any address owned by the exchange can hold both ETH and any of the ERC20. The order of the asset sums in the array is just a convention. Are you sure that we need to exclude any addresses from ERC20 or ETH calculation? If the exchange owns them, it inevitably owns the ETH or any ERC20 minted to that address.

[the remark this answers is not in the document]
Comment (alxkzmn) [reply contracts/src/Summa.sol:72]: Another option could be to have a separate variable for total ETH sum and an array for all the ERC20s.

[the remark this answers is not in the document]
Comment (enricobottazzi) [reply contracts/src/Summa.sol:72]: You are right, I think there was a misunderstanding (maybe given by the fact that the review is pointing at lines 31, 32,33, while I was just commenting line 34...)

Absolutely for each of `cexAddresses` we should fetch their balance for each `erc20ContractAddresses` and `ETH`, aggregate this balance and compare it to `assetsSum`

My proposal was to merge `erc20ContractAddresses` and `ETH` into a single `assetContractAddress`. 

So for example if `assetContractAddress` is `["0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
        "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"]` the function will:
 
- for each address in `cexAddresses` fetch the erc20 balance at contract "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48", aggregate the sum to `erc20Balances[i]` and verify that this is greater or equal than the corresponding `assetsSum` 
- for each address in `cexAddresses` fetch the balance at contract "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee". This is not an erc20 contract but means ETH (only exception that we need to handle now, but I think that's how we will manage also non-native cu rrencies such as BTC)

**Review by enricobottazzi — APPROVED** [review]
LGTM

**Review by sifnoc — APPROVED** [review]

**Pull request merged**